File: src/hollywood/adapters/wga.py
from __future__ import annotations

import json
import re
import time
from collections.abc import Iterable, Sequence
from pathlib import Path
from typing import Any

from ..config import HollywoodSettings
from ..models import (
    CreditRow,
    EntityAliasRow,
    EntityKind,
    EntityRow,
    IngestOptions,
    NormalizedBundle,
    RawPayload,
    SourceDefinition,
    canonicalize_url,
    json_dumps,
    make_stable_id,
)
from ..storage import HollywoodStorage
from .base import BaseAdapter

SEARCH_INPUT_SELECTORS = (
    "#Filter_Keyword",
    "input[name='Filter.Keyword']",
    "input[placeholder='Search writer or project here...']",
)
SEARCH_TYPE_BUTTON_SELECTORS = ("#search-type", "button#search-type")
SEARCH_BUTTON_SELECTORS = ("#searchBtn", "button#searchBtn")
STARTS_WITH_ITEM_SELECTOR = "a.dropdown-item[data-value='2']"
CREDIT_COUNT_RE = re.compile(r"^(?P<count>\d+)\s+Credits?$")


class SelectorError(RuntimeError):
    """Raised when the WGA page shape does not match expected selectors."""


def writer_key(url: str) -> str:
    return make_stable_id("wga", canonicalize_url(url))


def normalize_prefixes(raw: str) -> list[str]:
    value = raw.strip()
    if not value:
        raise ValueError("prefixes must not be empty")
    prefixes = [part.strip() for part in value.split(",")] if "," in value else list(value)
    cleaned = [prefix for prefix in prefixes if prefix]
    if not cleaned:
        raise ValueError("prefixes must contain at least one non-empty prefix")
    return cleaned


def unique_profile_urls(urls: Iterable[str]) -> list[str]:
    return sorted({canonicalize_url(url) for url in urls})


def _selector_error(page: Any, description: str, selectors: Sequence[str]) -> SelectorError:
    selector_list = ", ".join(selectors)
    return SelectorError(
        f"Could not find {description} on {page.url!r} using selectors: {selector_list}"
    )


def _click(page: Any, selectors: Sequence[str], description: str) -> None:
    for selector in selectors:
        locator = page.locator(selector)
        for index in range(locator.count()):
            candidate = locator.nth(index)
            if candidate.is_visible():
                candidate.click()
                return
    raise _selector_error(page, description, selectors)


def _fill(page: Any, selectors: Sequence[str], value: str, description: str) -> None:
    for selector in selectors:
        locator = page.locator(selector)
        for index in range(locator.count()):
            candidate = locator.nth(index)
            if candidate.is_visible():
                candidate.fill(value)
                return
    raise _selector_error(page, description, selectors)


def search_prefix(page: Any, prefix: str) -> None:
    _click(page, SEARCH_TYPE_BUTTON_SELECTORS, "the search type dropdown")
    _click(page, (STARTS_WITH_ITEM_SELECTOR,), "the 'Starts With' option")
    _fill(page, SEARCH_INPUT_SELECTORS, prefix, "the search input")
    _click(page, SEARCH_BUTTON_SELECTORS, "the 'Search' control")


def collect_profile_urls(page: Any) -> list[str]:
    urls = page.locator("a[href*='/member/']").evaluate_all("els => els.map(a => a.href)")
    return unique_profile_urls(urls)


def _extract_writer_name(profile: Any) -> str:
    for selector in ("h1", "main h1", ".page-title h1"):
        locator = profile.locator(selector)
        if locator.count():
            text = locator.first.inner_text().strip()
            if text:
                return text
    title = profile.title().strip()
    return title.split("|")[0].strip() if title else "Unknown Writer"


def _parse_wga_credits(text: str, person_entity_id: str) -> list[CreditRow]:
    credits: list[CreditRow] = []
    previous_title: str | None = None
    for line in [line.strip() for line in text.splitlines() if line.strip()]:
        match = CREDIT_COUNT_RE.match(line)
        if match and previous_title:
            count = int(match.group("count"))
            credits.append(
                CreditRow(
                    credit_id=make_stable_id("wga", person_entity_id, previous_title, str(count)),
                    source_id="wga",
                    person_entity_id=person_entity_id,
                    person_name=None,
                    title_name=previous_title,
                    title_external_id=None,
                    role="writer_credit_summary",
                    billing=count,
                    metadata_json=json_dumps({"source": "wga_profile_text"}),
                )
            )
            previous_title = None
            continue
        if (
            not line.startswith("Created by:")
            and not line.startswith("Written by:")
            and "Writers Guild" not in line
            and "Jump To" not in line
            and "(" in line
        ):
            previous_title = line
    return credits


class WgaAdapter(BaseAdapter):
    def __init__(self, source: SourceDefinition):
        super().__init__(source)

    def fetch_raw_payloads(
        self, settings: HollywoodSettings, options: IngestOptions
    ) -> list[RawPayload]:
        try:
            from playwright.sync_api import sync_playwright
        except ImportError as exc:  # pragma: no cover
            raise RuntimeError("Playwright is required for the WGA source.") from exc

        payloads: list[RawPayload] = []
        seen: set[str] = set()
        prefixes = options.prefixes or normalize_prefixes(
            str(self.source.metadata.get("default_prefixes", "abcdefghijklmnopqrstuvwxyz"))
        )
        emitted = 0

        with sync_playwright() as playwright:
            browser = playwright.chromium.launch(headless=True)
            context = browser.new_context(user_agent=settings.user_agent)
            try:
                page = context.new_page()
                for prefix in prefixes:
                    if options.limit is not None and emitted >= options.limit:
                        break
                    page.goto(self.source.default_urls[0], wait_until="networkidle")
                    search_prefix(page, prefix)
                    page.wait_for_load_state("networkidle")
                    for url in collect_profile_urls(page):
                        if url in seen:
                            continue
                        seen.add(url)
                        profile = context.new_page()
                        try:
                            profile.goto(url, wait_until="networkidle")
                            html = profile.content()
                            text = profile.locator("body").inner_text()
                            name = _extract_writer_name(profile)
                        finally:
                            profile.close()
                        canonical_url = canonicalize_url(url)
                        writer_id = writer_key(canonical_url)
                        metadata = {
                            "profile_url": canonical_url,
                            "writer_id": writer_id,
                            "writer_name": name,
                            "prefix": prefix,
                        }
                        payloads.append(
                            RawPayload(
                                payload_type="browser_html",
                                logical_id=writer_id,
                                body=html.encode("utf-8"),
                                content_type="text/html",
                                source_url=canonical_url,
                                canonical_url=canonical_url,
                                metadata=metadata,
                                extension=".html",
                            )
                        )
                        payloads.append(
                            RawPayload(
                                payload_type="browser_text",
                                logical_id=writer_id,
                                body=text.encode("utf-8"),
                                content_type="text/plain",
                                source_url=canonical_url,
                                canonical_url=canonical_url,
                                metadata=metadata,
                                extension=".txt",
                            )
                        )
                        emitted += 1
                        if options.limit is not None and emitted >= options.limit:
                            break
                        time.sleep(2)
            finally:
                context.close()
                browser.close()
        return payloads

    def normalize_raw_records(
        self,
        settings: HollywoodSettings,
        storage: HollywoodStorage,
        run_id: str,
        raw_records: list[dict[str, object]],
    ) -> NormalizedBundle:
        bundle = NormalizedBundle()
        for record in raw_records:
            if str(record["payload_type"]) != "browser_text":
                continue
            path = Path(str(record["content_path"]))
            metadata = json.loads(str(record["metadata_json"]))
            text = path.read_text(encoding="utf-8", errors="replace")
            writer_name = metadata.get("writer_name", "Unknown Writer")
            writer_id = metadata.get(
                "writer_id", writer_key(str(record["canonical_url"] or record["source_url"]))
            )
            entity_id = make_stable_id("wga", EntityKind.PERSON.value, str(writer_id))
            bundle.entities.append(
                EntityRow(
                    entity_id=entity_id,
                    source_id=self.source.source_id,
                    external_id=str(writer_id),
                    entity_type=EntityKind.PERSON.value,
                    name=str(writer_name),
                    canonical_name=str(writer_name).casefold(),
                    license_class=self.source.license_class.value,
                    metadata_json=json_dumps(
                        {
                            "profile_url": metadata.get("profile_url"),
                            "raw_record_id": record["id"],
                        }
                    ),
                )
            )
            bundle.entity_aliases.append(
                EntityAliasRow(
                    entity_alias_id=make_stable_id(entity_id, str(writer_name)),
                    entity_id=entity_id,
                    source_id=self.source.source_id,
                    alias=str(writer_name),
                )
            )
            bundle.credits.extend(_parse_wga_credits(text, entity_id))
        return bundle

File: src/ingestion/adapters/wga.ts
import { readFileSync } from 'node:fs';

import { chromium } from 'playwright';
import type { Locator, Page } from 'playwright';

import type { RawRecordRow } from '../../domain/repositories/RawRecordRepository.js';
import { env } from '../../env.js';
import { canonicalizeUrl, emptyBundle, makeStableId } from '../models.js';
import type {
  CreditRow,
  EntityAliasRow,
  EntityRow,
  IngestOptions,
  NormalizedBundle,
  RawPayload,
  SourceDefinition,
} from '../models.js';
import type { Adapter } from './base.js';
import { parseWgaProfileCredits } from './wga-credit-parser.js';

const SEARCH_INPUT_SELECTORS = [
  '#Filter_Keyword',
  "input[name='Filter.Keyword']",
  "input[placeholder='Search writer or project here...']",
];
const SEARCH_TYPE_BUTTON_SELECTORS = ['#search-type', 'button#search-type'];
const SEARCH_BUTTON_SELECTORS = ['#searchBtn', 'button#searchBtn'];
const STARTS_WITH_ITEM_SELECTOR = "a.dropdown-item[data-value='2']";

class SelectorError extends Error {}

function writerKey(url: string): string {
  return makeStableId('wga', canonicalizeUrl(url));
}

export function normalizePrefixes(raw: string): string[] {
  const value = raw.trim();
  if (!value) throw new Error('prefixes must not be empty');
  const prefixes = value.includes(',') ? value.split(',').map((p) => p.trim()) : value.split('');
  const cleaned = prefixes.filter(Boolean);
  if (!cleaned.length) throw new Error('prefixes must contain at least one non-empty prefix');
  return cleaned;
}

function uniqueProfileUrls(urls: string[]): string[] {
  return [...new Set(urls.map(canonicalizeUrl))].sort();
}

async function clickFirstVisible(
  page: Page,
  selectors: string[],
  description: string,
): Promise<void> {
  for (const selector of selectors) {
    const locator = page.locator(selector);
    const count = await locator.count();
    for (let i = 0; i < count; i++) {
      const candidate = locator.nth(i);
      if (await candidate.isVisible()) {
        await candidate.click();
        return;
      }
    }
  }
  throw new SelectorError(
    `Could not find ${description} on ${page.url()} using selectors: ${selectors.join(', ')}`,
  );
}

async function fillFirstVisible(
  page: Page,
  selectors: string[],
  value: string,
  description: string,
): Promise<void> {
  for (const selector of selectors) {
    const locator = page.locator(selector);
    const count = await locator.count();
    for (let i = 0; i < count; i++) {
      const candidate = locator.nth(i);
      if (await candidate.isVisible()) {
        await candidate.fill(value);
        return;
      }
    }
  }
  throw new SelectorError(
    `Could not find ${description} on ${page.url()} using selectors: ${selectors.join(', ')}`,
  );
}

async function searchPrefix(page: Page, prefix: string): Promise<void> {
  await clickFirstVisible(page, SEARCH_TYPE_BUTTON_SELECTORS, 'the search type dropdown');
  await clickFirstVisible(page, [STARTS_WITH_ITEM_SELECTOR], "the 'Starts With' option");
  await fillFirstVisible(page, SEARCH_INPUT_SELECTORS, prefix, 'the search input');
  await clickFirstVisible(page, SEARCH_BUTTON_SELECTORS, "the 'Search' control");
}

async function collectProfileUrls(page: Page): Promise<string[]> {
  const urls = await page
    .locator("a[href*='/member/']")
    .evaluateAll((els) => els.map((a) => (a as HTMLAnchorElement).href));
  return uniqueProfileUrls(urls);
}

async function extractWriterName(profile: Page): Promise<string> {
  // The live WGA directory renders the writer's name in an <h3> under
  // #main-member's header (verified against a real profile page); the <h1>
  // selectors below are the site's older/generic page-banner heading
  // ("Find A Writer") and are kept only as a fallback in case the markup
  // reverts, so they must be tried after the more specific selector.
  for (const selector of ['#main-member h3', '.header h3', 'h1', 'main h1', '.page-title h1']) {
    const locator = profile.locator(selector);
    if (await locator.count()) {
      const text = (await locator.first().innerText()).trim();
      if (text) return text;
    }
  }
  const title = (await profile.title()).trim();
  if (!title) return 'Unknown Writer';
  return title.split(/\s*[|-]\s*/)[0]!.trim() || 'Unknown Writer';
}

/** Maps parsed WGA credits (pure text parsing lives in wga-credit-parser.ts) into bundle rows. */
export function buildWgaCreditRows(
  text: string,
  personEntityId: string,
): { credit: CreditRow; title: EntityRow }[] {
  return parseWgaProfileCredits(text).map((parsed) => {
    const titleEntityId = makeStableId('wga_title', parsed.title);
    return {
      title: {
        entityId: titleEntityId,
        sourceId: 'wga',
        entityType: 'title',
        name: parsed.title,
        canonicalName: parsed.title.toLowerCase(),
        metadataJson: JSON.stringify({ stub: true, category: parsed.category }),
        titleType: parsed.category ?? undefined,
      },
      credit: {
        creditId: makeStableId('wga', personEntityId, parsed.title, parsed.role),
        sourceId: 'wga',
        personEntityId,
        titleEntityId,
        role: parsed.role,
        creditCategory: 'crew',
        billing: parsed.count,
        metadataJson: JSON.stringify({ source: 'wga_profile_text' }),
      },
    };
  });
}

export class WgaAdapter implements Adapter {
  constructor(public source: SourceDefinition) {}

  async fetchRawPayloads(options: IngestOptions): Promise<RawPayload[]> {
    const payloads: RawPayload[] = [];
    const seen = new Set<string>();
    const prefixes =
      options.prefixes ??
      normalizePrefixes(
        String(this.source.metadata['default_prefixes'] ?? 'abcdefghijklmnopqrstuvwxyz'),
      );
    let emitted = 0;

    const browser = await chromium.launch({ headless: true });
    try {
      const context = await browser.newContext({ userAgent: env.HOLLYWOOD_USER_AGENT });
      try {
        const page = await context.newPage();
        for (const prefix of prefixes) {
          if (options.limit !== undefined && emitted >= options.limit) break;
          await page.goto(this.source.defaultUrls[0]!, { waitUntil: 'networkidle' });
          await searchPrefix(page, prefix);
          await page.waitForLoadState('networkidle');

          for (const url of await collectProfileUrls(page)) {
            if (seen.has(url)) continue;
            seen.add(url);

            const profile = await context.newPage();
            let html: string;
            let text: string;
            let name: string;
            try {
              await profile.goto(url, { waitUntil: 'networkidle' });
              html = await profile.content();
              text = await profile.locator('body').innerText();
              name = await extractWriterName(profile);
            } finally {
              await profile.close();
            }

            const canonicalUrl = canonicalizeUrl(url);
            const writerId = writerKey(canonicalUrl);
            const metadata = {
              profileUrl: canonicalUrl,
              writerId,
              writerName: name,
              prefix,
            };

            payloads.push({
              payloadType: 'browser_html',
              logicalId: writerId,
              body: Buffer.from(html, 'utf-8'),
              contentType: 'text/html',
              sourceUrl: canonicalUrl,
              canonicalUrl,
              fetchedAt: new Date(),
              metadata,
              extension: '.html',
            });
            payloads.push({
              payloadType: 'browser_text',
              logicalId: writerId,
              body: Buffer.from(text, 'utf-8'),
              contentType: 'text/plain',
              sourceUrl: canonicalUrl,
              canonicalUrl,
              fetchedAt: new Date(),
              metadata,
              extension: '.txt',
            });

            emitted++;
            if (options.limit !== undefined && emitted >= options.limit) break;
            await new Promise((resolve) => setTimeout(resolve, 2000));
          }
        }
      } finally {
        await context.close();
      }
    } finally {
      await browser.close();
    }

    return payloads;
  }

  async normalizeRawRecords(_runId: string, rawRecords: RawRecordRow[]): Promise<NormalizedBundle> {
    const bundle = emptyBundle();
    for (const record of rawRecords) {
      if (record.payloadType !== 'browser_text') continue;
      const path = record.contentPath;
      const metadata = JSON.parse(record.metadataJson ?? '{}');
      const text = readFileSync(path, 'utf-8');
      const writerName = metadata.writerName ?? 'Unknown Writer';
      const writerId =
        metadata.writerId ?? writerKey(record.canonicalUrl ?? record.sourceUrl ?? '');
      const entityId = makeStableId('wga', 'person', String(writerId));

      const entityRow: EntityRow = {
        entityId,
        sourceId: this.source.sourceId,
        externalId: String(writerId),
        entityType: 'person',
        name: String(writerName),
        canonicalName: String(writerName).toLowerCase(),
        metadataJson: JSON.stringify({
          profileUrl: metadata.profileUrl ?? null,
          rawRecordId: record.id,
        }),
      };
      bundle.entities.push(entityRow);

      const aliasRow: EntityAliasRow = {
        entityAliasId: makeStableId(entityId, String(writerName)),
        entityId,
        sourceId: this.source.sourceId,
        alias: String(writerName),
      };
      bundle.entityAliases.push(aliasRow);

      for (const { credit, title } of buildWgaCreditRows(text, entityId)) {
        bundle.entities.push(title);
        bundle.credits.push(credit);
      }
    }
    return bundle;
  }
}
